include/zephyr/sys/util.h

```diff
@@ -110,6 +110,45 @@ extern "C" {
 
 #endif /* __cplusplus */
 
+/**
+ * @brief Whether @p ptr is an element of @p array
+ *
+ * This macro can be seen as a slightly stricter version of @ref PART_OF_ARRAY
+ * in that it also ensures that @p ptr is aligned to an array-element boundary
+ * of @p array.
+ *
+ * In C, passing a pointer as @p array causes a compile error.
+ *
+ * @param array the array in question
+ * @param ptr the pointer to check
+ *
+ * @return 1 if @p ptr is part of @p array, 0 otherwise
+ */
+#define IS_ARRAY_ELEMENT(array, ptr)                                                               \
+	((uint8_t *)&(array)[0] <= (uint8_t *)(ptr) &&                                             \
+	 (uint8_t *)(ptr) < (uint8_t *)&(array)[ARRAY_SIZE(array)] &&                              \
+	 ((uint8_t *)(ptr) - (uint8_t *)(array)) % sizeof((array)[0]) == 0)
+
+/**
+ * @brief Index of @p ptr within @p array
+ *
+ * With `CONFIG_ASSERT=y`, this macro will trigger a runtime assertion
+ * when @p ptr is not aligned to an array-element boundary of @p array.
+ *
+ * In C, passing a pointer as @p array causes a compile error.
+ *
+ * @param array the array in question
+ * @param ptr pointer to an element of @p array
+ *
+ * @return the array index of @p ptr within @p array, on success
+ * @return SIZE_MAX if @p ptr is not a valid element of @p array
+ */
+#define ARRAY_INDEX(array, ptr)                                                                    \
+	({                                                                                         \
+		__ASSERT_NO_MSG(IS_ARRAY_ELEMENT(array, ptr));                                     \
+		IS_ARRAY_ELEMENT(array, ptr) ? (typeof((array)[0]) *)(ptr) - (array) : SIZE_MAX;   \
```

Comment:
I'm torn about this. I force users to incur a performance hit with `IS_ARRAY_ELEMENT()` being called outside of an assertion. In that case, the second line should just be

```cpp
(typeof((array)[0]) *)(ptr) - (array);
```

Is it safe to assume that devs use `CONFIG_ASSERT=y` during bringup?